src/main/kotlin/app/batch/HBaseWriter.kt

```diff
@@ -18,46 +15,79 @@ import java.io.BufferedReader
+
     override fun write(items: MutableList<out DecompressedStream>) {
         items.forEach {
+            logger.info("Processing '${it.fileName}'.")
             val fileName = it.fileName
-            val filenamePattern = """(?<database>[a-z-]+)\.(?<collection>[a-z-]+)\.\d+\.json\.gz\.enc$"""
+            val filenamePattern = """(?<database>[a-z0-9-]+)\.(?<collection>[a-z0-9-]+)\.\d+\.json\.gz\.enc$"""
             val filenameRegex = Regex(filenamePattern, RegexOption.IGNORE_CASE)
             val matchResult = filenameRegex.find(fileName)
             if (matchResult != null) {
                 val groups = matchResult.groups
                 val database = groups[1]!!.value // can assert nun-null as it matched on the regex
                 val collection = groups[2]!!.value // ditto
                 val dataKeyResult: DataKeyResult = getDataKey(fileName)
-                val reader = BufferedReader(InputStreamReader(it.inputStream))
-                var line: String? = null
-                var lineNo = 1;
-                while ({ line = reader.readLine(); line }() != null) {
-                    lineNo++
-                    try {
-                        val parser: Parser = Parser.default()
-                        val stringBuilder = StringBuilder(line)
-                        val json = parser.parse(stringBuilder) as JsonObject
-                        val id = json.obj("_id")?.toJsonString()
-                        if (StringUtils.isNotBlank(id)) {
-                            val encryptionResult = encryptDbObject(dataKeyResult, line!!, fileName, id)
-                            logger.debug("Success '$fileName' line ${lineNo}.")
-                            val message = MessageProducer().produceMessage(json, encryptionResult, dataKeyResult,
-                                                                            database, collection)
-                            logger.info("Message: '$message'.")
+
+                BufferedReader(InputStreamReader(it.inputStream)).use { reader ->
+                    var line: String? = null
+                    var lineNo = 0;
+                    while ({ line = reader.readLine(); line }() != null) {
+                        lineNo++
+                        try {
+                            val json = messageUtils.parseJson(line)
+                            val id = messageUtils.getId(json)?.toJsonString()
+                            if (StringUtils.isNotBlank(id)) {
+                                val encryptionResult = encryptDbObject(dataKeyResult, line!!, fileName, id)
+                                val message = messageProducer.produceMessage(json, encryptionResult, dataKeyResult,
+                                        database, collection)
+                                val lastModifiedTimestampStr = messageUtils.getLastModifiedTimestamp(json)
+                                if (StringUtils.isNotBlank(lastModifiedTimestampStr)) {
```

Comment:
We could reverse this `if` to be:
```
if (StringUtils.isBlank(id)) {
    logger.info("Skipping record $lineNo in the file $fileName due to absence of id")
    return@forEachLine
}
```
to remove a level of nesting. Though I am not sure on the validity or implementation of the `return@...` since I haven't seen it before. `continue` is not implemented for lambda's in Kotlin :(